yjit/src/codegen.rs

```diff
@@ -3411,6 +3418,32 @@ fn jit_guard_known_klass(
             jit_chain_guard(JCC_JNE, jit, ctx, cb, ocb, max_chain_depth, side_exit);
             ctx.upgrade_opnd_type(insn_opnd, Type::Flonum);
         }
+    } else if unsafe { known_klass == rb_cString } && sample_instance.string_p() {
+        assert!(!val_type.is_imm());
+        if val_type != Type::String {
```

Comment:
I don't think this optimization is entirely valid 😅. `Type::String` guarantees that the object is a `T_STRING` but does NOT guarantee that it is a `rb_cString` (it can be any subclass of `String`).

The reason the other special-cases here can be done is that they correspond to immediates whose subclasses can't be instantiated (ie. you're allowed to `class Foo < Integer` but `Foo.new` won't work, but for String it will, and this is used in practice).

Right now Type mostly corresponds to what type (`T_*`) it is (or whether or not its immediate for the unknown types). I think to do this we'd need to extend `Type` to encode exactly what we know about an object: its type? its class? its value? Doing this _should_ allow us to do this generically, for more classes than just string. But until then I don't think we can use this special case.

Reply:
``` ruby
class MyString < String
  def to_s
    "meh"
  end
end

def test_to_s(obj)
  "" == obj
  obj.to_s
end

OBJ = MyString.new

2.times {
  test_to_s("")
}

2.times {
  p test_to_s(OBJ)
}
```

```
$ ./miniruby test.rb
"meh"
"meh"

$ ./miniruby --yjit-call-threshold=2 test.rb
"meh"
""
```

Reply:
You're right. That definitely looks like a genuine bug. And you're right, currently our context checks for "string or subclass of string," not for "specifically un-subclassed string," which makes a difference in a lot of places in CRuby.

And you're right, that test case demonstrates that bug. Okay. I think this will need to be reverted :-/

CC @maximecb